### PR TITLE
Don't throw exception on m.login.jwt automatic user creation

### PR DESCRIPTION
**Pull request by olof**
When a previously non-existent user is authenticated with m.login.jwt (after the token has been validated), the user should be created. But a bug made checking if the user exists reset the claimed username to None, and we would see generic TypeError exceptions when we asked synapse to
process it further:
    
      ...
      File "/usr/lib/python3/dist-packages/synapse/rest/client/v1/login.py", line 393, in do_jwt_login
        result = await self._complete_login(
      File "/usr/lib/python3/dist-packages/synapse/rest/client/v1/login.py", line 338, in _complete_login
        localpart=UserID.from_string(user_id).localpart
      File "/usr/lib/python3/dist-packages/synapse/types.py", line 171, in from_string
        if len(s) < 1 or s[0:1] != cls.SIGIL:
      TypeError: object of type 'NoneType' has no len()
    
This regression was introduced in 541f1b92d (part of the v1.6.0 release of synapse); it only affects auth flows where _complete_login is called with create_non_existant_user=True, which turns out is only the m.login.jwt auth flow.

This PR also includes a commit that changes the parameter `create_non_existant_user` to `create_non_existent_user`, because... proactive confusion avoidance?

### Pull Request Checklist
[✓] on all checkboxes, as far as I can judge, but first time contributor here, so review my change with diligence! :)

### CODE REVIEW

**Comment by olof** [comment]
Nope, botched up the changelog thing. I'm on it.

**Comment by olof** [comment]
Looking at the failed test, it errors with `could not serialize access due to concurrent update`; doesn't look related to this change. Is that a reasonable conclusion?

**Comment by clokep** [comment]
> Looking at the failed test, it errors with `could not serialize access due to concurrent update`; doesn't look related to this change. Is that a reasonable conclusion?

I kicked off the test run again and it seems to have passed. 😄

File: changelog.d/7585.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+- Fix bug in automatic user creation on first time login with m.login.jwt. Regression in v1.6.0.
```
Comment (clokep) [inline changelog.d/7585.bugfix:1]: The list markup gets added automatically by the tooling. I made a couple of suggestions below too:

```suggestion
Fix a bug in automatic user creation during first time login with `m.login.jwt`. Regression in v1.6.0.
```

You can also include "Contributed by <your name or handle>." if you'd like.

Comment (olof) [reply changelog.d/7585.bugfix]: Thanks for the review! I've removed the list markup and incorporated your suggestions.

**Comment by olof** [comment]
I haven't looked at writing unit tests for this flow, no. Would love to give it a try. A new pull request for that would be better, right?

**Comment by olof** [comment]
I've written some unit tests, but as they are based on this change, I'm holding off on opening the pull request until this lands.

If you like to take a peek the tests are available at https://github.com/olof/synapse/tree/jwt_unit_tests

**Comment by clokep** [comment]
> A new pull request for that would be better, right?

You can just include them in this PR! No need to make it a separate one.

**Comment by olof** [comment]
Alright, tests added in this pull requests. Do I need to adapt the changelog, or that's ok as is?

**Comment by clokep** [comment]
The changelog looks good. I'll take another look at the code soon! 👍

**Comment by clokep** [comment]
Neat. This seem sane! I ran the tests without the fixes and two of the tests failed, so that's perfect. 👍

**Comment by clokep** [comment]
Also looks like #7604 got filed, which is essentially this same issue.

**Pull request merged**